common/src/main/java/io/netty/util/internal/ObjectPool.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2019 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.util.internal;
+
+import io.netty.util.Recycler;
+
+/**
+ * Light-weight object pool.
+ *
+ * @param <T> the type of the pooled object
+ */
+public abstract class ObjectPool<T> {
+
+    ObjectPool() { }
+
+    /**
+     * Get a {@link Object} from the {@link ObjectPool}. The returned {@link Object} may be created via
+     * {@link ObjectCreator#newObject(Handle)} if no pooled {@link Object} is ready to be reused.
+     */
+    public abstract T get();
+
+    /**
+     * Handle for an pooled {@link Object} that will be used to notify the {@link ObjectPool} once it can
+     * reuse the pooled {@link Object} again.
+     * @param <T>
+     */
+    public interface Handle<T> {
+        /**
+         * Recycle the {@link Object} if possible and so make it ready to be reused.
+         */
+        void recycle(T self);
+    }
+
+    /**
+     * Creates a new Object which references the given {@link Handle} and calls {@link Handle#recycle(Object)} once
+     * it can be re-used.
+     *
+     * @param <T> the type of the pooled object
+     */
+    public interface ObjectCreator<T> {
+
+        /**
+         * Creates an returns a new {@link Object} that can be used and later recycled via
+         * {@link Handle#recycle(Object)}.
+         */
+        T newObject(Handle<T> handle);
+    }
+
+    /**
+     * Creates a new {@link ObjectPool} which will use the given {@link ObjectCreator} to create the {@link Object}
+     * that should be pooled.
+     */
+    public static <T> ObjectPool<T> newPool(final ObjectCreator<T> creator) {
+        ObjectUtil.checkNotNull(creator, "creator");
+
+        return new RecyclerObjectPool<T>(creator);
```

Comment:
We will add more implementations here then which will be selected based on a system property.